experimental/ragged_inference/garbage_pad_ragged_acts.py

```diff
@@ -0,0 +1,191 @@
+
+    # Calculate the offsets for the padded acts
+    n_ctx_max = meta["n_ctx_max"]
+    padded_acts_offset = n_ctx_max * seq_idx * BLOCK_SIZE
+
+    # Write things back, again masking out the sections that would be garbage
+    tl.store(padded_acts_ptr + padded_acts_offset, acts, mask=ctx_idx_too_large_mask)
+
+
+class RaggedActivations:
+    def __init__(self, raw_tensor: torch.Tensor, n_ctx_per_seq: List[int]):
+        self.raw_tensor = raw_tensor
+        self.n_ctx_per_seq = n_ctx_per_seq
+
+    @property
+    def n_seqs(self):
+        return len(self.n_ctx_per_seq)
+
+    @property
+    def max_n_ctx_per_seq(self):
+        return max(self.n_ctx_per_seq)
+
+    @property
+    def dtype(self):
+        return self.raw_tensor.dtype
+
+    @property
+    def device(self):
+        return self.raw_tensor.device
+
+    @classmethod
+    def from_list(cls, tensors: List[torch.Tensor]):
+        """Tensors must all be of shape [n_ctx, d_model]."""
+        return cls(
+            raw_tensor=torch.cat(tensors),
+            n_ctx_per_seq=[tensor.shape[0] for tensor in tensors],
+        )
+
+    def iter_full_tensors(self):
+        idx_so_far = 0
+        for n_ctx_in_this_seq in self.n_ctx_per_seq:
+            yield self.raw_tensor[idx_so_far : idx_so_far + n_ctx_in_this_seq]
+            idx_so_far += n_ctx_in_this_seq
+
+    def to_garbage_padded(self) -> torch.Tensor:
+        """
+        Create a tensor of shape (n_seqs, n_ctx_max, d_model) where the
+        sequences are right-padded with garbage data
+        """
+        n_seqs = len(self.n_ctx_per_seq)
+        n_ctx_max = max(self.n_ctx_per_seq)
+
+        n_dim = self.raw_tensor.shape[-1]
+        # TODO: flag use zeros for garbage
+        padded_acts = torch.zeros(
+            n_seqs, n_ctx_max, n_dim, dtype=self.raw_tensor.dtype, device="cuda"
+        )
+
+        idx_so_far = 0
+        for seq_idx, n_ctx_in_this_seq in enumerate(self.n_ctx_per_seq):
+            this_seq = self.raw_tensor[idx_so_far : idx_so_far + n_ctx_in_this_seq]
+            padded_acts[seq_idx, :n_ctx_in_this_seq, :] = this_seq
+            idx_so_far += n_ctx_in_this_seq
+
+        return padded_acts
+
+    def triton_to_garbage_padded(self) -> torch.Tensor:
+        """
+        Create a tensor of shape (n_seqs, n_ctx_max, d_model) where the
+        sequences are right-padded with garbage data
+        """
+        n_seqs = len(self.n_ctx_per_seq)
+        n_ctx_max = max(self.n_ctx_per_seq)
+
+        ragged_acts = self.raw_tensor
+        d_model = ragged_acts.shape[-1]
+        padded_acts = torch.empty(
+            n_seqs, n_ctx_max, d_model, dtype=ragged_acts.dtype, device="cuda"
+        )
+
+        # We just use one program per n_ctx position for simplicity
+        assert d_model >= 128, f"bad {d_model=}"
+        assert d_model <= 8 * 1024, f"bad {d_model=}"
+        assert d_model % 32 == 0, f"bad {d_model=}"
```

Comment:
Why 32 and not 64? Where did these requirements come from?

Reply:
I think it comes from BLOCK_K = 32, a scheduling/tiling constraint for the matmuls. This could be relaxed if we were to mask over that dimension in the kernel